pkg/controller/kubelet-config/kubelet_config_bootstrap.go

```diff
@@ -0,0 +1,183 @@
+package kubeletconfig
+
+import (
+	"encoding/json"
+	"fmt"
+
+	ign3types "github.com/coreos/ignition/v2/config/v3_2/types"
+	"github.com/imdario/mergo"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	mtmpl "github.com/openshift/machine-config-operator/pkg/controller/template"
+	"github.com/openshift/machine-config-operator/pkg/version"
+	"github.com/vincent-petithory/dataurl"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+)
+
+// RunKubeletBootstrap generates MachineConfig objects for mcpPools that would have been generated by syncKubeletConfig
+// nolint:gocyclo
+func RunKubeletBootstrap(templateDir string, kubeletConfigs []*mcfgv1.KubeletConfig, controllerConfig *mcfgv1.ControllerConfig, mcpPools []*mcfgv1.MachineConfigPool) ([]*mcfgv1.MachineConfig, error) {
+	var (
+		kubeletIgnition            *ign3types.File
+		logLevelIgnition           *ign3types.File
+		autoSizingReservedIgnition *ign3types.File
+	)
+	userDefinedSystemReserved := make(map[string]string)
+
+	// Validate the KubeletConfig CR if exists
+	for _, kubeletConfig := range kubeletConfigs {
+		if kubeletConfig == nil {
+			continue
+		}
+		if err := validateUserKubeletConfig(kubeletConfig); err != nil {
+			return nil, newForgetError(err)
+		}
+	}
+
+	var res []*mcfgv1.MachineConfig
+
+	for _, kubeletConfig := range kubeletConfigs {
+		if kubeletConfig == nil || kubeletConfig.Spec == (mcfgv1.KubeletConfigSpec{}) {
+			continue
+		}
+		for _, pool := range mcpPools {
+			// skip the pool if no default label for kubeletconfig
+			_, hasMasterLabel := pool.Labels["pools.operator.machineconfiguration.openshift.io/master"]
+			_, hasWorkerLabel := pool.Labels["pools.operator.machineconfiguration.openshift.io/worker"]
+			if !hasMasterLabel && !hasWorkerLabel {
+				continue
+			}
+			role := pool.Name
+			managedKey, err := getManagedKeyKubelet(pool, nil)
```

Comment:
So, if I'm interpreting this correctly, the managedKey is always `99-%s-%s-kubelet` and not using the incremental schema like the main KCC (https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/common/constants.go#L20).

This means that if I define multiple KCs for a pool, the later one that gets into this loop would take priority, which I think unfortunately drifts us again between the main KCC and the bootstrap one.

Since we are being very restrictive (only allowing pool labelling), we could go one step further and say you can only define 1 KCC per pool (max 2 bootstrap KCCs: master/worker since you can't use custom labels). What do you think about that? Or we can beef this up a bit to use the same Suffix schema as the main KCC

Reply:
I would prefer to share the main KCC code to stay the same imlementation, but the result will be the same as the latter kubeletconfig takes the priority, since currently, new kubeltconfig will ocerride any changes from the previous one?

Reply:
I think so (should double check to make sure but I don't think it does a merge), but the difference is: in the main KCC there is 1 MC per KC, whereas here there would only be 1 MC total (for the last KC).

Drift aside, if the user ever deletes the "newer" kubeletconfig, we may lose the old one as well since there is no MC tracking it.

Reply:
I think this might be hitting on my [comment below](https://github.com/openshift/machine-config-operator/pull/2547#discussion_r678688555)(correct me if I'm wrong). 

It feels like a user should come out of an install with the cleanest slate possible and also minimizing the potential errors just after bringing up a cluster and as @yuqi-zhang points out, maybe unable to recover from normal behavior. 

While the code might stay the same, it sounds like the functional underpinnings will be different, and I think we need to make sure that we end up with a safe state so that a user is able to conduct normal functions without damaging the cluster. I dont think ending up with multiple kubelet configs is very ergonomic in a new cluster, but if they also don't all have tracking MCs (as they would in a day2 KC) this doesn't seem like a route we should be taking.

Reply:
I think Qi has since updated the PR to port over the "in-cluster KCC" mode of operation. For the comment below, this does allow us to have multiple KCs (up to the regular in-cluster limit of 10), so if that works we have parity with the "in-cluster MCC". I think if we can test out that functionality to be the same between bootstrap + in-cluster KCC, this would be the "safest" route. 

I agree that having multiple KCs is probably not a good idea during bootstrap, but as long as our logic can account for the multiple-per-pool use case I think we shouldn't run into any trouble